plugins/sharding/window.rb

```diff
@@ -0,0 +1,236 @@
+          next_edge = Time.at(current_min.to_time.to_i + @step_second)
+          year = next_edge.year
+          month = next_edge.month
+          day = next_edge.day
+          if @unit == :day
+            hour = 0
+          else
+            hour = next_edge.hour
+          end
+          minute = 0
+          second = 0
+          microsecond = 0
+          border = :include
+          WindowEdge.new(year, month, day, hour, minute, second, microsecond,
+                         border)
+      end
+    
+      def select_by_range(table, shard_key, min, max)
+        expression = Expression.create(table)
+        begin
+          expression.append_object(@between, Operator::PUSH, 1)
+          expression.append_object(shard_key, Operator::PUSH, 1)
+          expression.append_operator(Operator::GET_VALUE, 1)
+          expression.append_constant(min.to_s, Operator::PUSH, 1)
+          expression.append_constant(min.border, Operator::PUSH, 1)
+          expression.append_constant(max.to_s, Operator::PUSH, 1)
+          expression.append_constant(max.border, Operator::PUSH, 1)
+          expression.append_operator(Operator::CALL, 5)
+          table.select(expression)
+        ensure
+          expression.close
+        end
+      end
+    
+      def inspect_range(min, max)
+        range = ""
+        if min.border == :include
+          range << "["
+        else
+          range << "("
+        end
+        range << min.to_s
+        range << ","
+        range << max.to_s
+        if max.border == :include
+          range << "]"
+        else
+          range << ")"
+        end
+        range
+      end
+    end
+    
+    class WindowEdge
+      include Comparable
+      
+      attr_reader :year
+      attr_reader :month
+      attr_reader :day
+      attr_reader :hour
+      attr_reader :minute
+      attr_reader :second
+      attr_reader :microsecond
+      attr_reader :border
+      def initialize(year, month, day, hour, minute, second, microsecond,
+                     border)
+        @year = year
+        @month = month
+        @day = day
+        @hour = hour
+        @minute = minute
+        @second = second
+        @microsecond = microsecond
+        @border = border
+      end
+      
+      def to_s
+        format = "%04d/%02d/%02d %02d:%02d:%02d"
+        format_values = [@year, @month, @day, @hour, @minute, @second]
+        unless @microsecond.zero?
+          format << ".%06d"
+          format_values << @microsecond
+        end
+        format % format_values
+      end
+      
+      def to_time
+        Time.local(@year, @month, @day, @hour, @minute, @second, @microsecond)
+      end
+      
+      def values
+        [@year, @month, @day, @hour, @minute, @second, @microsecond]
+      end
+      
+      def <=>(other)
+        (values + [@border == :include ? 1 : 0]) <=>
+          (other.values + [other.border == :include ? 1 : 0])
+      end
+    end
+  end
+end
```

Comment:
Could you add a new line?

Reply:
Sorry for forgetting it so many times...
I have fixed it.